iree/compiler/Dialect/Flow/Transforms/MatmulToMMT4dConversion.cpp

```diff
@@ -0,0 +1,200 @@
+                                       int K0, PatternBenefit benefit = 1)
+      : OpRewritePattern<linalg::MatmulOp>(context, benefit),
+        M0Size(M0),
+        N0Size(N0),
+        K0Size(K0) {}
+
+  LogicalResult matchAndRewrite(linalg::MatmulOp matmulOp,
+                                PatternRewriter &rewriter) const override {
+    auto loc = matmulOp.getLoc();
+
+    Value lhs = matmulOp.getInputOperand(0)->get();
+    Value rhs = matmulOp.getInputOperand(1)->get();
+    Value dst = matmulOp.getOutputOperand(0)->get();
+
+    RankedTensorType lhsType = lhs.getType().dyn_cast<RankedTensorType>();
+    RankedTensorType rhsType = rhs.getType().dyn_cast<RankedTensorType>();
+    RankedTensorType dstType = dst.getType().dyn_cast<RankedTensorType>();
+
+    if (!lhsType || !rhsType || !lhsType.hasStaticShape() ||
+        !rhsType.hasStaticShape()) {
+      return failure();
+    }
+
+    // This is for float only matmul for now. Integer data type might require
+    // r.h.s layout change.
+    if (!lhsType.getElementType().isF32() || !rhsType.getElementType().isF32())
+      return failure();
+
+    int m = lhsType.getShape()[0];
+    int n = rhsType.getShape()[1];
+    int k = rhsType.getShape()[0];
+
+    if (m % M0Size != 0 || n % N0Size != 0 || k % K0Size != 0) return failure();
+
+    int m1 = m / M0Size;
+    int n1 = n / N0Size;
+    int k1 = k / K0Size;
+
+    // Expands a 2d tensor operand to 4d given its target shape.
+    auto expandTo4D = [&](Value operand,
+                          ArrayRef<int64_t> targetShape) -> Value {
+      auto operandType = operand.getType().cast<RankedTensorType>();
+      auto targetType =
+          RankedTensorType::get(targetShape, operandType.getElementType());
+      SmallVector<ReassociationIndices> expandIndices = {{0, 1}, {2, 3}};
+      Value reshapedOperand = rewriter.create<linalg::TensorExpandShapeOp>(
+          loc, targetType, operand, expandIndices);
+      return reshapedOperand;
+    };
+
+    bool isInitTensorDst = false;
+    if (dst.getDefiningOp() && isa<linalg::FillOp>(dst.getDefiningOp())) {
+      auto fillOpValue = cast<linalg::FillOp>(dst.getDefiningOp());
+      if (fillOpValue.output().getDefiningOp() &&
+          isa<linalg::InitTensorOp>(fillOpValue.output().getDefiningOp())) {
+        isInitTensorDst = true;
+      }
+    }
+    auto createTransposedInitTensor =
+        [&](ArrayRef<int64_t> targetShape) -> Value {
+      return rewriter.create<linalg::InitTensorOp>(loc, targetShape,
+                                                   dstType.getElementType());
+    };
+
+    auto lhs4D = expandTo4D(lhs, {m1, M0Size, k1, K0Size});
+    auto rhs4D = expandTo4D(rhs, {k1, K0Size, n1, N0Size});
+    auto dst4D = isInitTensorDst
+                     ? createTransposedInitTensor({m1, n1, M0Size, N0Size})
+                     : expandTo4D(dst, {m1, M0Size, n1, N0Size});
+    auto transposeOperand = [&](Value operand,
```

Comment:
Nit: Not a big proponent of huge lambdas like this. Move this into a separate `static ` function outside of the function? Also add some comments as to what this is doing.

Reply:
I am so used to lambdas, but you do more code health work here so I will go with your suggestion :)